crowbar_framework/app/models/service_object.rb

```diff
@@ -1540,6 +1524,79 @@ def wait_for_chef_daemons(node_list)
 
   private
 
+  THREAD_POOL_SIZE = 20
+
+  def release_chef_locks(locks)
+    return if locks.empty?
+
+    queue = Queue.new
+    locks.each { |l| queue.push l }
+
+    workers = (1...THREAD_POOL_SIZE).map do
+      Thread.new do
+        loop do
+          begin
+            lock = queue.pop(true)
+          rescue ThreadError
+            break
+          end
+
+          lock.release
+        end
+      end
+    end
+
+    logger.debug "Waiting for #{THREAD_POOL_SIZE} unlock threads to finish..."
+    workers.map(&:join)
+  end
+
+  def lock_and_wait_for_chef_clients(nodes, lock_owner, lock_reason)
+    locks = []
+    errors = {}
+
+    return [locks, errors] if nodes.empty?
+
+    locks_mutex = Mutex.new
+    errors_mutex = Mutex.new
+
+    queue = Queue.new
+    nodes.each { |n| queue.push n }
```

Comment:
It better be, as it seems that its only purpouse its to sync between threads :)

http://ruby-doc.org/stdlib-2.0.0/libdoc/thread/rdoc/Queue.html